### ISSUE TITLE
Docs: consistent-type-imports vs importsNotUsedAsValues

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#6393]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

The `@typescript-eslint/consistent-type-imports` rule flags the following code:

```ts
import { MyType } from "./MyType";
```

And forces you to change it to:

```ts
import type { MyType } from "./MyType";
```

However, it seems that the [`importsNotUsedAsValues`](https://www.typescriptlang.org/tsconfig#importsNotUsedAsValues) compiler flag does the same thing (when set to the "error" value).

Thus, the documentation for the `@typescript-eslint/consistent-type-imports` rule should provide a short discussion of the differences between the two, if any, and why you would want to use one over the other.

In issue #4268, Brad explains that you should definitely not use both of them at the same time, but fails to mention anything further.

### Affected URL(s)

https://typescript-eslint.io/rules/consistent-type-imports/

**Comment by bradzacher** [comment]
Note: in TS 5.0 `importsNotUsedAsValues` is be deprecated in favour of the new `verbatimModuleSyntax` flag.

---

> but fails to mention anything further

The comment you mentioned was the reasoning behind the closing sentence.
I.e. you could reorder the comment as this:

> You should either use the TS option, or our lint rule - not both.
> Because ... 

---

In most cases you can use them both (if for some reason you want double reporting).
The only time they will conflict is if you use experimental decorators AND decorator metadata because (as the comment mentioned) TS does some type-aware, cross-module codegen that causes implicit references.

**Comment by Zamiell** [comment]
Thanks Brad.

It is my understanding that importing for side effects is a Very Bad Thing. So, I would imagine most TypeScript users wouldn't care about the new behavior you mention with `verbatimModuleSyntax`.

If the TypeScript compiler flag catches decorator stuff, and the ESLint rule doesn't, then why does the rule exist at all? Should the documentation for the rule point people towards disabling the rule and enabling the compiler flag?

Thinking about this more, I do prefer to use ESLint rules to TypeScript compiler flags sometimes. Specifically, I disable `--noUnusedLocals` in favor of `@typescript-eslint/no-unused-vars`, because it has more functionality - you can ignore arbitrary variable prefixes. (The TypeScript compiler offers no such customizability.)

But this logic doesn't seem to apply for the `importsNotUsedAsValues`/`verbatimModuleSyntax` case. The ESLint rule just seems strictly worse in every scenario!

**Comment by bradzacher** [comment]
> then why does the rule exist at all

1) a lot of people don't know about the compiler option - it was a minor footnote in the TS3.8 release notes that added the top-level `import type` syntax.
   Most people at best skim read the release notes, so it was easy to miss. Heck I even missed it and I have to write up the release notes for our repo here.
1) a lot of people don't like these kinds of "linty" checks being done by TS, doubly so if they need to turn on a specific flag.
1) for a codebase of non-trivial size - doing a non-type-aware lint is faster than doing a `tsc` check over the entire codebase, so you can get faster feedback from CI and such with lint rules.
1) eslint rules have autofixers which can be applied in bulk with `--fix` or applied on save in the IDE. TS fixers cannot have either done.
1) you have to set `importsNotUsedAsValues: 'error'` to get TS to report on the imports.
   This is bad for a lot of people because it `'error'` implies the `'preserve'` option - meaning TS will leave behind value specifiers (and entire import declarations) that are unused (which can lead to the aforementioned runtime side effects).
1) the vast majority of users DO NOT use decorators, let alone decorators with metadata. So they won't ever run into the inconsistency.
   I think angular is one of the few frameworks that relies on decorator metadata?
1) [this rule will also ban inline type imports by default](https://typescript-eslint.io/rules/consistent-type-imports/#disallowtypeannotations), which TS has no check to do.

The big one really is (5) - using `importsNotUsedAsValues`/`verbatimModuleSyntax` will cause TS to emit different runtime code. You cannot have TS do these checks without it also changing the runtime code. In conjunction with (6) - most people never run into any inconsistencies.

**Comment by Zamiell** [comment]
Fantastic, thanks Brad.

We should definitely incorporate all of the info in Brad's last two comments in [the webpage for the "consistent-type-imports" rule](https://typescript-eslint.io/rules/consistent-type-imports/).

**Comment by bradzacher** [comment]
Just to demonstrate the side effects thing:

For the following code:
```ts
import { type ComponentClass } from 'react';

export type T = ComponentClass;
```

This is the resulting JS code for different compiler options:

### Default options (which is `importsNotUsedAsValues: remove` / `verbatimModuleSyntax: false`)

```ts
export {};
```
[playground](https://www.typescriptlang.org/play?importsNotUsedAsValues=0#code/PTAEAEEsFsAcHsBOAXAzgOXsgqqgpgCYCCqAagIYA2ArnqgFyh6KJICwAUJzAiqAN6hkAT1h5QAYXhx4AOzyzkEyuVSpQAX1AAzVtFAByRHnIBjZAYDcnTngAevZENHiAKqAC8k6QnmLlqqjWHEA)

### <TS5.0 - `importsNotUsedAsValues: error`

```ts
import 'react';
```
[playground](https://www.typescriptlang.org/play?importsNotUsedAsValues=2#code/PTAEAEEsFsAcHsBOAXAzgOXsgqqgpgCYCCqAagIYA2ArnqgFyh6KJICwAUJzAiqAN6hkAT1h5QAYXhx4AOzyzkEyuVSpQAX1AAzVtFAByRHnIBjZAYDcnTngAevZENHiAKqAC8k6QnmLlqqjWHEA)

### >= TS5.0 - `verbatimModuleSyntax: true`

```ts
import {} from 'react';
```
[playground](https://www.typescriptlang.org/play?verbatimModuleSyntax=true&ts=5.0.0-dev.20230130#code/PTAEAEDcFMCcCMCGAXAlgWwLIHsAmBXAG2gGUBPAO2UQA8AuUZWfaAWACgOMAHbWZUAG9GZbtFABhbOl4VoVCYUQBnZaAC+oAGaxpoAOSxoiAMbJ9Abg4doNXvxFjQAFVABeSdNnzkilcqt2IA)

---

This clearly illustrates why using the lint rule + default option is what a lot of people prefer - because it means you don't accidentally leave behind runtime side effects for what looks like a type-only import.
